File: clients/ts/src/wire.ts
/**
 * The `find_route` wire: every number this package encodes, in one place.
 *
 * The instruction data is fixed-width: a 20-byte header, then one 4-byte menu entry per pool, and
 * nothing else — the program refuses data of any other length. The header is the discriminator,
 * the flags byte, `max_walk_steps`, `num_mints`, `num_pools`, and `min_profit_base_units` as a
 * little-endian `u64`. A menu entry is `hop_kind`, `account_count`, and the two transfer-hook group
 * lengths, which this package always sends as zero.
 *
 * Each constant here is held against `wire/wire-manifest.json` by the test suite. A wire literal
 * anywhere else in the package is a defect.
 */
import { address, type Address } from "@solana/addresses";
import {
  fixEncoderSize,
  type FixedSizeEncoder,
  type ReadonlyUint8Array,
} from "@solana/codecs-core";
import { getArrayEncoder, getBytesEncoder, getStructEncoder } from "@solana/codecs-data-structures";
import { getU64Encoder, getU8Encoder } from "@solana/codecs-numbers";

import { FindRouteError } from "./error.js";

/**
 * The wire revision this package encodes for. A consumer pins it; a mismatch against the deployed
 * program means the two disagree about the instruction's shape.
 */
export const WIRE_EPOCH = 2;

export type RouterProgramAddress = "TURKbNaes5RA3sMnkRsCmuPBKZbPTyRxv9cTiMQ43Am";

/** The router program. */
export const ROUTER_PROGRAM_ADDRESS: Address<RouterProgramAddress> = address(
  "TURKbNaes5RA3sMnkRsCmuPBKZbPTyRxv9cTiMQ43Am",
);

/** The one seed of the router's config account. */
export const CONFIG_SEED = "config";

/** The instruction discriminator: `sha256("global:find_route")[..8]`. */
export const FIND_ROUTE_DISC: ReadonlyUint8Array = new Uint8Array([
  0x63, 0x61, 0x70, 0x5d, 0x8f, 0x05, 0x5e, 0x00,
]);

/** Bytes before the first menu entry. */
export const HEADER_LEN = 20;

/** Bytes per menu entry. */
export const MENU_ENTRY_LEN = 4;

/** The most pools a menu may name. */
export const MAX_MENU_POOLS = 8;

/** The most route mints an instruction may name. */
export const MAX_ROUTE_MINTS = 4;

/** The ceiling on every window's `account_count`, summed over the menu. */
export const MAX_MENU_ACCOUNTS = 69;

/**
 * The ceiling on one transfer-hook account group. This package declares no groups; the number is
 * carried so the manifest agreement is complete.
 */
export const MAX_HOOK_GROUP_LEN = 21;

/**
 * Flags bit 0: the base token account holds borrowed principal a later instruction repays. The
 * program then behaves as if {@link FLAG_FAIL_IF_NO_PROFIT} were set.
 */
export const FLAG_FLASHLOAN = 1;

/**
 * Flags bit 1: fail the instruction, rather than settle a loss, when no cycle clears the profit
 * threshold.
 */
export const FLAG_FAIL_IF_NO_PROFIT = 2;

/** Wrapped SOL, the first base mint. */
export const WSOL_MINT_ADDRESS: Address<"So11111111111111111111111111111111111111112"> = address(
  "So11111111111111111111111111111111111111112",
);

/** USDC, the second base mint. */
export const USDC_MINT_ADDRESS: Address<"EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"> = address(
  "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
);

export interface PrefixAccountMeta {
  readonly name: string;
  readonly isSigner: boolean;
  readonly isWritable: boolean;
}

/** The six accounts before the route-mint section, with the flags a caller sends. */
export const PREFIX_ACCOUNT_METAS: readonly [
  PrefixAccountMeta,
  PrefixAccountMeta,
  PrefixAccountMeta,
  PrefixAccountMeta,
  PrefixAccountMeta,
  PrefixAccountMeta,
] = [
  { name: "user", isSigner: true, isWritable: false },
  { name: "base_ata", isSigner: false, isWritable: true },
  { name: "base_mint", isSigner: false, isWritable: false },
  { name: "base_token_program", isSigner: false, isWritable: false },
  { name: "config_account", isSigner: false, isWritable: false },
  { name: "fee_ata", isSigner: false, isWritable: true },
];

/** The header fields a caller chooses. Counts are supplied by the builder, which has bounded them. */
export interface FindRouteHeader {
  readonly flags: number;
  readonly maxWalkSteps: number;
  readonly numMints: number;
  readonly numPools: number;
  readonly minProfitBaseUnits: bigint;
}

/** One menu entry: the four bytes `[hop_kind, account_count, hook_lens[0], hook_lens[1]]`. */
export interface MenuEntry {
  readonly hopKind: number;
  readonly accountCount: number;
  readonly hookLen0: number;
  readonly hookLen1: number;
}

interface HeaderWithDiscriminator extends FindRouteHeader {
  readonly discriminator: ReadonlyUint8Array;
}

/** @internal */
export function getFindRouteHeaderEncoder(): FixedSizeEncoder<HeaderWithDiscriminator> {
  return getStructEncoder([
    ["discriminator", fixEncoderSize(getBytesEncoder(), FIND_ROUTE_DISC.length)],
    ["flags", getU8Encoder()],
    ["maxWalkSteps", getU8Encoder()],
    ["numMints", getU8Encoder()],
    ["numPools", getU8Encoder()],
    ["minProfitBaseUnits", getU64Encoder()],
  ]);
}

/** @internal */
export function getMenuEntryEncoder(): FixedSizeEncoder<MenuEntry> {
  return getStructEncoder([
    ["hopKind", getU8Encoder()],
    ["accountCount", getU8Encoder()],
    ["hookLen0", getU8Encoder()],
    ["hookLen1", getU8Encoder()],
  ]);
}

/**
 * Lays out the header and the entries; exactly `HEADER_LEN + MENU_ENTRY_LEN * entries.length`
 * bytes.
 * @internal
 */
export function encodeFindRouteData(
  header: FindRouteHeader,
  entries: ReadonlyArray<MenuEntry>,
): ReadonlyUint8Array {
  const headerBytes = getFindRouteHeaderEncoder().encode({
    discriminator: FIND_ROUTE_DISC,
    ...header,
  });
  const entryBytes = getArrayEncoder(getMenuEntryEncoder(), { size: entries.length }).encode([
    ...entries,
  ]);
  const data = new Uint8Array(headerBytes.length + entryBytes.length);
  data.set(headerBytes, 0);
  data.set(entryBytes, headerBytes.length);
  return data;
}

const U8_MAX = 255n;
const U64_MAX = 0xffff_ffff_ffff_ffffn;

/**
 * A byte the wire carries as given. Refused by name before any encoder runs, so a JavaScript
 * caller's out-of-range number is never silently wrapped.
 * @internal
 */
export function assertU8(field: "maxWalkSteps", value: number): void {
  if (!Number.isInteger(value) || value < 0 || value > 255) {
    throw new FindRouteError({
      kind: "IntegerOutOfRange",
      field,
      given: value,
      min: 0n,
      max: U8_MAX,
    });
  }
}

/** @internal */
export function assertU64(field: "minProfitBaseUnits", value: bigint): void {
  if (typeof value !== "bigint" || value < 0n || value > U64_MAX) {
    throw new FindRouteError({
      kind: "IntegerOutOfRange",
      field,
      given: value,
      min: 0n,
      max: U64_MAX,
    });
  }
}

File: clients/rust/turk-router/src/wire.rs
//! The `find_route` wire: every number this crate encodes, in one place.
//!
//! The instruction data is fixed-width: a 20-byte header, then one 4-byte menu entry per pool,
//! and nothing else — the program refuses data of any other length, padded or truncated. The
//! header is the discriminator, the flags byte, `max_walk_steps`, `num_mints`, `num_pools`, and
//! `min_profit_base_units` as a little-endian `u64`. A menu entry is `hop_kind`, `account_count`,
//! and the two transfer-hook group lengths, which this crate always sends as zero.
//!
//! Each constant here is held against `wire/wire-manifest.json` by the test suite. A wire literal
//! anywhere else in the crate is a defect.

use solana_pubkey::Pubkey;

/// `TURKbNaes5RA3sMnkRsCmuPBKZbPTyRxv9cTiMQ43Am` — the router program.
pub const ROUTER_PROGRAM_ID: Pubkey = Pubkey::new_from_array([
    6, 200, 26, 170, 58, 47, 218, 193, 91, 166, 109, 28, 87, 105, 91, 122, 172, 249, 66, 243, 185,
    234, 214, 247, 216, 205, 127, 253, 82, 0, 152, 38,
]);

/// The one seed of the router's config account.
pub const CONFIG_SEED: &[u8] = b"config";

/// The instruction discriminator: `sha256("global:find_route")[..8]`.
pub const FIND_ROUTE_DISC: [u8; 8] = [0x63, 0x61, 0x70, 0x5d, 0x8f, 0x05, 0x5e, 0x00];

/// Bytes before the first menu entry.
pub const HEADER_LEN: usize = 20;

/// Bytes per menu entry.
pub const MENU_ENTRY_LEN: usize = 4;

/// The most pools a menu may name.
pub const MAX_MENU_POOLS: usize = 8;

/// The most route mints an instruction may name.
pub const MAX_ROUTE_MINTS: usize = 4;

/// The ceiling on every window's `account_count`, summed over the menu.
pub const MAX_MENU_ACCOUNTS: usize = 69;

/// The ceiling on one transfer-hook account group. This crate declares no groups; the number is
/// carried so the manifest agreement is complete.
pub const MAX_HOOK_GROUP_LEN: usize = 21;

/// Flags bit 0: the base token account holds borrowed principal a later instruction repays. The
/// program then behaves as if [`FLAG_FAIL_IF_NO_PROFIT`] were set.
pub const FLAG_FLASHLOAN: u8 = 1 << 0;

/// Flags bit 1: fail the instruction, rather than settle a loss, when no cycle clears the profit
/// threshold.
pub const FLAG_FAIL_IF_NO_PROFIT: u8 = 1 << 1;

/// The six accounts before the route-mint section, with the flags a caller sends:
/// `(name, is_signer, is_writable)`.
pub const PREFIX_ACCOUNT_METAS: [(&str, bool, bool); 6] = [
    ("user", true, false),
    ("base_ata", false, true),
    ("base_mint", false, false),
    ("base_token_program", false, false),
    ("config_account", false, false),
    ("fee_ata", false, true),
];

/// The header fields a caller chooses. Counts are supplied by the builder, which has already
/// bounded them.
pub(crate) struct Header {
    pub(crate) flags: u8,
    pub(crate) max_walk_steps: u8,
    pub(crate) num_mints: u8,
    pub(crate) num_pools: u8,
    pub(crate) min_profit_base_units: u64,
}

/// Lays out the header and the entries; exactly `HEADER_LEN + MENU_ENTRY_LEN * entries.len()`
/// bytes.
pub(crate) fn encode(header: &Header, entries: &[[u8; MENU_ENTRY_LEN]]) -> Vec<u8> {
    let mut data =
        Vec::with_capacity(HEADER_LEN.saturating_add(entries.len().saturating_mul(MENU_ENTRY_LEN)));
    data.extend_from_slice(&FIND_ROUTE_DISC);
    data.push(header.flags);
    data.push(header.max_walk_steps);
    data.push(header.num_mints);
    data.push(header.num_pools);
    data.extend_from_slice(&header.min_profit_base_units.to_le_bytes());
    for entry in entries {
        data.extend_from_slice(entry);
    }
    data
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_program_id_is_the_documented_address() {
        assert_eq!(
            ROUTER_PROGRAM_ID.to_string(),
            "TURKbNaes5RA3sMnkRsCmuPBKZbPTyRxv9cTiMQ43Am"
        );
    }

    #[test]
    fn the_header_is_twenty_bytes_in_field_order() {
        let header = Header {
            flags: FLAG_FAIL_IF_NO_PROFIT,
            max_walk_steps: 3,
            num_mints: 2,
            num_pools: 1,
            min_profit_base_units: 0x0102_0304_0506_0708,
        };
        let data = encode(&header, &[[0, 9, 0, 0]]);
        assert_eq!(data.len(), HEADER_LEN + MENU_ENTRY_LEN);
        assert_eq!(&data[..8], &FIND_ROUTE_DISC);
        assert_eq!(&data[8..12], &[2, 3, 2, 1]);
        assert_eq!(&data[12..20], &[8, 7, 6, 5, 4, 3, 2, 1]);
        assert_eq!(&data[20..], &[0, 9, 0, 0]);
    }
}
